### PR TITLE
feat: support dash in args (#4519)

### PR DESCRIPTION
**Pull request by nanamikon**
### What this PR does / why we need it:

Support fetch value from arg name with dash , see #4519

### Pre-submission checklist:

* [ 1] Did you explain what problem does this PR solve? Or what new features have been added?
* [ 1] Have you added corresponding test cases?
* [ 1] Have you modified the corresponding document?
* [ 1] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

File: t/core/ctx.t [hunk 1]
```diff
@@ -1008,3 +1008,23 @@ GET /hello
 hello world
 --- error_log eval
 qr/request log: \{"route_id":"1","route_name":"my_route","service_id":"1","service_name":"my_service"\}/
+
+
+
+=== TEST 36: support dash in the args
```
Comment (spacewander) [inline t/core/ctx.t:1014]: Could you add the test in t/core/ctx2.t? Note that we already set up request & no_error_log there:
https://github.com/apache/apisix/blob/e3e018315311e0ca201f17579f9106eb88a834d1/t/core/ctx2.t#L23-L37

File: apisix/core/ctx.lua [hunk 2]
```diff
@@ -159,10 +160,22 @@ do
                     val, err = cookie:get(sub_str(key, 8))
                     if err then
                         log.warn("failed to fetch cookie value by key: ",
-                                 key, " error: ", err)
+                                key, " error: ", err)
```
Comment (tokers) [inline apisix/core/ctx.lua:163]: Please revert this change, the original indent is good.

File: apisix/core/ctx.lua [hunk 3]
```diff
@@ -159,10 +160,22 @@ do
...
                     end
                 end
 
+            elseif core_str.has_prefix(key, "arg_") then
+                local args = t.args
+                local arg_key = sub_str(key, 5);
```
Comment (tokers) [inline apisix/core/ctx.lua:169]:
```suggestion
                local arg_key = sub_str(key, 5)
```

File: apisix/core/ctx.lua [hunk 4]
```diff
@@ -159,10 +160,22 @@ do
...
+                local m, err = ngx_re_match(args, "(^|&)" .. arg_key .. "=([^&]*)(&|$)", "jo")
+                if err then
+                    log.warn("failed to fetch arg value by key: ",
+                            key, " error: ", err)
```
Comment (tokers) [inline apisix/core/ctx.lua:173]:
```suggestion
                    log.warn("failed to fetch arg value by key: ",
                             key, " error: ", err)
```

File: t/core/ctx.t (hunk 1 above)
Comment (nanamikon) [reply t/core/ctx.t:1014]: ok

File: apisix/core/ctx.lua (hunk 2 above)
Comment (nanamikon) [reply apisix/core/ctx.lua:163]: ok

File: apisix/core/ctx.lua [hunk 5]
```diff
@@ -159,10 +160,22 @@ do
...
                         log.warn("failed to fetch cookie value by key: ",
-                                 key, " error: ", err)
+                                key, " error: ", err)
                     end
                 end
 
+            elseif core_str.has_prefix(key, "arg_") then
+                local args = t.args
+                local arg_key = sub_str(key, 5)
+                local m, err = ngx_re_match(args, "(^|&)" .. arg_key .. "=([^&]*)(&|$)", "jo")
```
Comment (spacewander) [inline apisix/core/ctx.lua:170]: We can use `core.request.get_uri_args`, which is more clear?

Comment (nanamikon) [reply apisix/core/ctx.lua:170]: Do you mean?
```
args :　a-b=ab
core.request.get_uri_args_a-b :  ab
```

if  *uri*  is for making args different from args in nginx var，  I think the following example is clear enough?
```
args :　a-b=ab
core.request.get_args_a-b :  ab
```

Comment (spacewander) [reply apisix/core/ctx.lua:170]: Something like `core.request.get_uri_args()["a-b"]`, note that if the result is a table, get the first one.

Comment (nanamikon) [reply apisix/core/ctx.lua:170]:
```
                        {
                            "vars": [
                                ["uri_args_a-b","==","ab"]
                            ]
                        }
```
If get_uri_args return a table,  wil this work properly?

Comment (spacewander) [reply apisix/core/ctx.lua:170]: Note that if the result is a table, get the first one.

Comment (nanamikon) [reply apisix/core/ctx.lua:170]: get it

File: apisix/core/ctx.lua [hunk 6]
```diff
@@ -164,16 +163,15 @@ do
...
-                local m, err = ngx_re_match(args, "(^|&)" .. arg_key .. "=([^&]*)(&|$)", "jo")
-                if err then
-                    log.warn("failed to fetch arg value by key: ",
-                             key, " error: ", err)
-                end
-                if m then
-                    val = m[2]
+            elseif core_str.has_prefix(key, "uri_args_") then
+                local arg_key = sub_str(key, 10)
+                local args = ngx.req.get_uri_args()[arg_key]
```
Comment (spacewander) [inline apisix/core/ctx.lua:168]: Let's use `core.request.get_uri_args`

File: apisix/core/ctx.lua [hunk 7]
```diff
@@ -164,16 +163,15 @@ do
...
-                local args = t.args
-                local arg_key = sub_str(key, 5)
-                local m, err = ngx_re_match(args, "(^|&)" .. arg_key .. "=([^&]*)(&|$)", "jo")
-                if err then
-                    log.warn("failed to fetch arg value by key: ",
-                             key, " error: ", err)
-                end
-                if m then
-                    val = m[2]
+            elseif core_str.has_prefix(key, "uri_args_") then
```
Comment (spacewander) [inline apisix/core/ctx.lua:166]: Err... Let's use the old name "arg_".

Comment (spacewander) [reply apisix/core/ctx.lua:166]: Sorry for my mistake, I didn't make it clear

File: t/core/ctx2.t [hunk 8]
```diff
@@ -136,3 +136,39 @@ qr/balancer_ip: 127.0.0.1|balancer_port: 1980/
...
+        }
+    }
+--- request
+GET /t?a-b=aaa
+--- response_body
+uri_args_a-b: aaa
+
+
+
+=== TEST 6: support dash in the args(Multi args with the same name, only fetch the first one)
```
Comment (spacewander) [inline t/core/ctx2.t:160]: Let's add a test that arg is missing

Comment (nanamikon) [reply t/core/ctx2.t:160]: ok

File: apisix/core/ctx.lua (hunk 7 above)
Comment (nanamikon) [reply apisix/core/ctx.lua]: It's ok

File: apisix/core/ctx.lua [hunk 9]
```diff
@@ -159,7 +159,18 @@ do
...
                     if err then
                         log.warn("failed to fetch cookie value by key: ",
-                                 key, " error: ", err)
+                                key, " error: ", err)
+                    end
+                end
+
+            elseif core_str.has_prefix(key, "arg_") then
+                local arg_key = sub_str(key, 5)
+                local args = request.get_uri_args(ngx.ctx)[arg_key]
```
Comment (spacewander) [inline apisix/core/ctx.lua:168]: We don't need to pass `ngx.ctx`

Comment (nanamikon) [reply apisix/core/ctx.lua:168]: I add ngx.ctx, because the following exception
```
 Failed test 'TEST 7: support dash in the args(arg is missing) - pattern "[error]" should not match any line in error.log but matches line "2021/07/30 18:02:36 [error] 17776\#574167: *10 lua entry thread aborted: runtime error: /Users/jianzhou/workspace/apisix/apisix/core/request.lua:132: attempt to index local 'ctx' (a nil value)" (req 0)
# stack traceback:
# coroutine 0:
#       /Users/jianzhou/workspace/apisix/apisix/core/request.lua: in function 'get_uri_args'
#       /Users/jianzhou/workspace/apisix/apisix/core/ctx.lua:168: in function '__index'
#       content_by_lua(nginx.conf:179):6: in main chunk, client: 127.0.0.1, server: localhost, request: "GET /t HTTP/1.1", host: "localhost"

```

Comment (spacewander) [reply apisix/core/ctx.lua:168]: The api_ctx is set by the other part.
You can refer to this test: https://github.com/apache/apisix/pull/4723/files

Comment (nanamikon) [reply apisix/core/ctx.lua]: Get it,   I find it inited in the init.lua

File: t/core/ctx2.t [hunk 10]
```diff
@@ -136,3 +136,69 @@ qr/balancer_ip: 127.0.0.1|balancer_port: 1980/
...
+
+=== TEST 5: support dash in the args
+--- config
+    location /t {
+        content_by_lua_block {
+            local core = require("apisix.core")
+            local ngx_ctx = ngx.ctx
+            local api_ctx = ngx_ctx.api_ctx
+            if api_ctx == nil then
+                api_ctx = core.tablepool.fetch("api_ctx", 0, 32)
```
Comment (spacewander) [inline t/core/ctx2.t:150]: It is not a good idea to set up api_ctx directly. We should let APISIX set it. See https://github.com/apache/apisix/pull/4676#discussion_r680302823

File: apisix/core/ctx.lua (hunk 2 above)
Comment (tzssangglass) [reply apisix/core/ctx.lua:162]: bad indent

**Comment by spacewander** [comment]
Please merge master to make CI pass.

**Comment by nanamikon** [comment]
Some conflicts should be solved,   i will fix it

**Comment by nanamikon** [comment]
> Please merge master to make CI pass.

How to deal with these ci test error?   It seems not relate to my pr

**Comment by nanamikon** [comment]
> Please merge master to make CI pass.

CI fail refering to error log
```
2021/08/03 22:55:24 [error] 21865#948770: *10 lua entry thread aborted: runtime error: /Users/jianzhou/workspace/apisix/apisix/core/request.lua:132: attempt to index local 'ctx' (a nil value)
stack traceback:
coroutine 0:
        /Users/jianzhou/workspace/apisix/apisix/core/request.lua: in function 'get_uri_args'
        /Users/jianzhou/workspace/apisix/apisix/core/ctx.lua:169: in function '__index'
        /Users/jianzhou/workspace/apisix/apisix/admin/init.lua:71: in function 'check_token'
        /Users/jianzhou/workspace/apisix/apisix/admin/init.lua:123: in function 'handler'
        ...workspace/apisix//deps/share/lua/5.1/resty/radixtree.lua:722: in function 'dispatch'
        /Users/jianzhou/workspace/apisix/apisix/init.lua:765: in function 'http_admin'
        content_by_lua(nginx.conf:222):2: in main chunk, client: 127.0.0.1, server: , request: "GET /apisix/admin/routes HTTP/1.1", host: "127.0.0.1:9180"

```

I think it should pass api_ctx to ngx.ctx in the /admin/init.lua
```
local function run()
    local api_ctx = {}
    core.ctx.set_vars_meta(api_ctx)

    local ok, err = check_token(api_ctx)
```

Do I need to fit this in this pr?

**Comment by spacewander** [comment]
@nanamikon 
Yes, let's do it.

**Comment by nanamikon** [comment]
All tests pass now

**Comment by spacewander** [comment]
We can remove the 'arg' part from 'TEST 2: http header + arg'

**Comment by nanamikon** [comment]
Get it

**Comment by spacewander** [comment]
Err... We still need the "TEST 2: http header" but remove the arg part of it.

**Comment by nanamikon** [comment]
> Err... We still need the "TEST 2: http header" but remove the arg part of it.

Sorry for misunderstanding， I will fix it

File: apisix/core/ctx.lua (hunk 2 above)
Comment (tzssangglass) [reply apisix/core/ctx.lua:163]:
```suggestion
                                 key, " error: ", err)
```

**Review by spacewander — APPROVED** [review]

**Review by membphis — APPROVED** [review]
